streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceivedBlockTracker.scala

```diff
@@ -195,31 +285,56 @@ private[streaming] class ReceivedBlockTracker(
       timeToAllocatedBlocks --= batchTimes
     }
 
+    def resolveEvent(event: ReceivedBlockTrackerLogEvent): Unit = {
+      event match {
+        case CombinedRBTLogEvent(events) => events.foreach(resolveEvent)
+        case BlockAdditionEvent(receivedBlockInfo) =>
+          insertAddedBlock(receivedBlockInfo)
+        case BatchAllocationEvent(time, allocatedBlocks) =>
+          insertAllocatedBatch(time, allocatedBlocks)
+        case BatchCleanupEvent(batchTimes) =>
+          cleanupBatches(batchTimes)
+      }
+    }
+
     writeAheadLogOption.foreach { writeAheadLog =>
       logInfo(s"Recovering from write ahead logs in ${checkpointDirOption.get}")
       writeAheadLog.readAll().asScala.foreach { byteBuffer =>
         logTrace("Recovering record " + byteBuffer)
-        Utils.deserialize[ReceivedBlockTrackerLogEvent](
-          byteBuffer.array, Thread.currentThread().getContextClassLoader) match {
-          case BlockAdditionEvent(receivedBlockInfo) =>
-            insertAddedBlock(receivedBlockInfo)
-          case BatchAllocationEvent(time, allocatedBlocks) =>
-            insertAllocatedBatch(time, allocatedBlocks)
-          case BatchCleanupEvent(batchTimes) =>
-            cleanupBatches(batchTimes)
-        }
+        resolveEvent(Utils.deserialize[ReceivedBlockTrackerLogEvent](
+          byteBuffer.array, Thread.currentThread().getContextClassLoader))
       }
     }
   }
 
   /** Write an update to the tracker to the write ahead log */
-  private def writeToLog(record: ReceivedBlockTrackerLogEvent) {
+  private def writeToLog(record: ReceivedBlockTrackerLogEvent): Boolean = {
     if (isWriteAheadLogEnabled) {
-      logDebug(s"Writing to log $record")
       writeAheadLogOption.foreach { logManager =>
+        logTrace(s"Writing record: $record")
         logManager.write(ByteBuffer.wrap(Utils.serialize(record)), clock.getTimeMillis())
       }
     }
+    true
+  }
+
+  import WALWriteStatus._
+
+  /**
+   * Adds LogEvents to a queue so that they can be batched and written to the WAL.
+   * Exposed for testing.
+   */
+  private[streaming] def writeToLogAsync(event: ReceivedBlockTrackerLogEvent): Boolean = {
```

Comment:
`writeToLogAsync` looks confusing. Actually, `async` usually means putting the work in another thread and return immediately, and it should run pretty fast. But in this method, it actually needs to wait for the result.